web/server/vue-cli/src/views/OAuthLogin.vue

```diff
@@ -0,0 +1,124 @@
+<template>
+  <v-container>
+    <v-row>
+      <v-col cols="12">
+        <v-card>
+          <v-card-text class="px-0 pb-0">
+            <alerts
+              :success="success"
+              success-msg="Successfully logged in!"
+              :error="error"
+              :error-msg="errorMsg"
+            />
+          </v-card-text>
+        </v-card>
+      </v-col>
+    </v-row>
+  </v-container>
+</template>
+
+<script>
+import { mapGetters } from "vuex";
+import { LOGIN } from "@/store/actions.type";
+import Alerts from "@/components/Alerts.vue";
+
+
+export default {
+  name: "OAuthLogin",
+  components: {
+    alerts: Alerts
+  },
+
+  data() {
+    return {
+      success: false,
+      error: false,
+      errorMsg: null,
+    };
+  },
+
+  computed: {
+    ...mapGetters([
+      "isAuthenticated"
+    ])
+  },
+
+  created() {
+    if (this.isAuthenticated) {
+      this.$router.replace({ name: "products" });
+    }
+  },
+
+  mounted() {
+    this.detectCallback();
+  },
+
+  methods: {
+    detectCallback() {
+      const url = this.$route.query;
+      const provider = this.$route.params.provider;
+      const state = localStorage.getItem("oauth_state");
+      const code_challenge = localStorage.getItem("code_challenge");
+      const method = localStorage.getItem("method");
+
+      if (url.code != null && url.state != null) {
+        if (url.state != state) {
+          this.errorMsg = "Invalid state!";
+          this.error = true;
+          this.callback = true;
+          return;
+        }
+
+        const oauth_data_id = localStorage.getItem("oauth_data_id");
+        const baseUrl = window.location.href.replace("#", "");
+        const baseUrlOAuthId = `${baseUrl}&oauth_data_id=${oauth_data_id}`;
+        const baseMethod = `${baseUrlOAuthId}&code_challenge_method=${method}`;
+        const fullUrl = `${baseMethod}&code_challenge=${code_challenge}`;
```

Comment:
This snippet looks suspect. Why are you reconstructing a URL yourself based on what's in the localStorage, when you are trying to pass the callback URL along as-is? In fact, why are these details stored at all, other than what is strictly necessary?

The stored details in localStorage, and all other variables are checked on the server side, and that is the only check that matters in the chain.

These details in the localStorage also become prone to the same thing that caused the mix-up attack - when someone dispatches a second OAuth login session, and they still didn't finish the first one, these stored details are overwritten.

Reply:
fixed

Reply:
fixed, The localStorage is not used anymore, the necessary arguments are stored on server side in database, the oauth_id is removed, the database querying is happening based on state ,